## North Star Metric

It is the single metric that best captures the core value that your product delivers to customers.

This metric depends on your company’s product, position, targets & more. Airbnb’s North Star Metric is nights booked whereas for Facebook, it is daily active users.

[Dataset](https://www.kaggle.com/vijayuv/onlineretail) : Sample dataset of an online retail store

In [74]:

# import libraries
from datetime import datetime, timedelta
import pandas as pd
%matplotlib inline
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
#from __future__ import division

import plotly as py
import plotly.offline as pyoff
import plotly.graph_objs as go

#initiate visualization library for jupyter notebook 
#pyoff.init_notebook_mode()

tx_data = pd.read_csv('data/OnlineRetail.csv', encoding = 'unicode_escape')

tx_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

Revenue = Active Customer Count * Order Count * Average Revenue per Order

In [75]:

#converting the type of Invoice Date Field from string to datetime.
tx_data['InvoiceDate'] = pd.to_datetime(tx_data['InvoiceDate'])


#converting the type of Invoice Date Field from string to datetime.
tx_data['InvoiceYearMonth'] = tx_data['InvoiceDate'].map(lambda date: 100*date.year + date.month)

#compute revenue for each row
tx_data['Revenue'] = tx_data['Quantity'] * tx_data['UnitPrice']

In [76]:
tx_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   

   InvoiceYearMonth  Revenue  
0            201012    15.30  
1            201012    20.34  
2            201012    22.00  
3            201012    20.34  
4            201012    20.34

In [77]:
monthly_revenue = tx_data.groupby(by='InvoiceYearMonth')['Revenue'].sum().reset_index()
monthly_revenue

InvoiceYearMonth      Revenue
0             201012   748957.020
1             201101   560000.260
2             201102   498062.650
3             201103   683267.080
4             201104   493207.121
5             201105   723333.510
6             201106   691123.120
7             201107   681300.111
8             201108   682680.510
9             201109  1019687.622
10            201110  1070704.670
11            201111  1461756.250
12            201112   433668.010

In [78]:
plot_data = [
    go.Scatter(
    x=monthly_revenue['InvoiceYearMonth'],
    y=monthly_revenue['Revenue']
    )
]

plot_layout = go.Layout(
    xaxis={"type":"category"},
    title='Monthly Revenue'
)

fig = go.Figure(data=plot_data, layout=plot_layout)
fig.show()

This clearly shows our revenue is growing especially Aug ‘11 onwards (and our data in December is incomplete). Let’s figure out what is our **Monthly Revenue Growth Rate**:

In [79]:
#using pct_change() function to see monthly percentage change
monthly_revenue['MonthlyGrowth'] = monthly_revenue['Revenue'].pct_change()
monthly_revenue.head()

InvoiceYearMonth     Revenue  MonthlyGrowth
0            201012  748957.020            NaN
1            201101  560000.260      -0.252293
2            201102  498062.650      -0.110603
3            201103  683267.080       0.371850
4            201104  493207.121      -0.278163

In [87]:
plot_data = [
    go.Scatter(
        x=monthly_revenue.query("InvoiceYearMonth < 201112")['InvoiceYearMonth'],
        y=monthly_revenue.query("InvoiceYearMonth < 201112")['MonthlyGrowth'],
    )
]
plot_data

[Scatter({
     'x': array([201012, 201101, 201102, 201103, 201104, 201105, 201106, 201107, 201108,
                 201109, 201110, 201111], dtype=int64),
     'y': array([        nan, -0.2522932 , -0.11060282,  0.37184967, -0.2781635 ,
                  0.46659178, -0.04453048, -0.01421311,  0.00202612,  0.49365275,
                  0.05003204,  0.36522824])
 })]

In [88]:
#visualization - line graph
plot_data = [
    go.Scatter(
        x=monthly_revenue.query("InvoiceYearMonth < 201112")['InvoiceYearMonth'],
        y=monthly_revenue.query("InvoiceYearMonth < 201112")['MonthlyGrowth'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Montly Growth Rate'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)

There is a huge drop in revenue in April. Was it due to less active customers or our customers did less orders? Maybe they just started to buy cheaper products? We can’t say anything without doing a deep-dive analysis.

### Monthly Active Customers
To see the details Monthly Active Customers, we will follow the steps we exactly did for Monthly Revenue. Starting from this part, we will be focusing on UK data only (which has the most records).

In [108]:
#filter for only uk customers
tx_uk = tx_data.query("Country=='United Kingdom'").reset_index(drop=True)
tx_monthly_active = tx_uk.groupby(by='InvoiceYearMonth')['CustomerID'].nunique().reset_index()
tx_monthly_active

InvoiceYearMonth  CustomerID
0             201012         871
1             201101         684
2             201102         714
3             201103         923
4             201104         817
5             201105         985
6             201106         943
7             201107         899
8             201108         867
9             201109        1177
10            201110        1285
11            201111        1548
12            201112         617

In [109]:
#visualization - bar graph
plot_data = [
    go.Bar(
        x=tx_monthly_active['InvoiceYearMonth'],
        y=tx_monthly_active['CustomerID'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Montly Active Customers'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)

In [105]:
(923-817)/923

0.11484290357529794

We see that in April 2011, there is a 11.5% drop in customers

### Monthly order count

In [110]:
tx_monthly_sales = tx_uk.groupby(by='InvoiceYearMonth')['Quantity'].sum().reset_index()
tx_monthly_sales

InvoiceYearMonth  Quantity
0             201012    298101
1             201101    237381
2             201102    225641
3             201103    279843
4             201104    257666
5             201105    306452
6             201106    258522
7             201107    324129
8             201108    319804
9             201109    458490
10            201110    455612
11            201111    642281
12            201112    199907

In [111]:
#plot
plot_data = [
    go.Bar(
        x=tx_monthly_sales['InvoiceYearMonth'],
        y=tx_monthly_sales['Quantity'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Montly Sales'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)

In [112]:
(279843-257666)/279843

0.07924800691816483

As expected, in April, there is a drop in sales as well. (8%) We can say that decline in number of active customers affected the order count.

### Average revenue per order

In [113]:
tx_uk.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   

   InvoiceYearMonth  Revenue  
0            201012    15.30  
1            201012    20.34  
2            201012    22.00  
3            201012    20.34  
4            201012    20.34

In [115]:
tx_monthly_order_avg = tx_uk.groupby(by='InvoiceYearMonth')['Revenue'].mean().reset_index()
tx_monthly_order_avg

InvoiceYearMonth    Revenue
0             201012  16.865860
1             201101  13.614680
2             201102  16.093027
3             201103  16.716166
4             201104  15.773380
5             201105  17.713823
6             201106  16.714748
7             201107  15.723497
8             201108  17.315899
9             201109  18.931723
10            201110  16.093582
11            201111  16.312383
12            201112  16.247406

In [116]:
#plot
plot_data = [
    go.Bar(
        x=tx_monthly_order_avg['InvoiceYearMonth'],
        y=tx_monthly_order_avg['Revenue'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly order Average'
    )

fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)

Even the monthly order average dropped for April (16.7 to 15.8). We observed slow-down in every metric affecting our North Star.

Let’s continue investigating some other important metrics:
* **New Customer Ratio**: a good indicator of if we are losing our existing customers or unable to attract new ones
* **Retention Rate**: King of the metrics. Indicates how many customers we retain over specific time window. We will be showing examples for monthly retention rate and cohort based retention rate.

### New Customer Ratio
First we should define what is a new customer. In our dataset, we can assume a new customer is whoever did his/her first purchase in the time window we defined. We will do it monthly for this example.

We will be using .min() function to find our first purchase date for each customer and define new customers based on that. The code below will apply this function and show us the revenue breakdown for each group monthly.

In [122]:
#create a dataframe contaning CustomerID and first purchase date
tx_first_purchase = tx_uk.groupby(by='CustomerID')['InvoiceDate'].min().reset_index()
tx_first_purchase.columns = ['CustomerID', 'FirstPurchase']
tx_first_purchase.head()

CustomerID       FirstPurchase
0     12346.0 2011-01-18 10:01:00
1     12747.0 2010-12-05 15:38:00
2     12748.0 2010-12-01 12:48:00
3     12749.0 2011-05-10 15:25:00
4     12820.0 2011-01-17 12:34:00

In [126]:
tx_first_purchase['FirstPurchaseYearMonth'] = tx_first_purchase['FirstPurchase'].map(lambda date: date.year*100 + date.month)
tx_first_purchase.head()

CustomerID       FirstPurchase  FirstPurchaseYearMonth
0     12346.0 2011-01-18 10:01:00                  201101
1     12747.0 2010-12-05 15:38:00                  201012
2     12748.0 2010-12-01 12:48:00                  201012
3     12749.0 2011-05-10 15:25:00                  201105
4     12820.0 2011-01-17 12:34:00                  201101

In [127]:
tx_uk = pd.merge(tx_uk, tx_first_purchase, on='CustomerID')
tx_uk.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom   

   InvoiceYearMonth  Revenue       FirstPurchase  FirstPurchaseYearMonth  
0            201012    15.30 2010-12-01 08:26:00                  201012  
1            201012    20.34 2010-12-01 08:26:00                  201012  
2            201012    22.00 2010-12-01 08:26:00                  201012  
3            201012    20.34 2010-12-01 08:26:00                  201012  
4            201012    20.34 2010-12-01 08:26:00                  201012

In [132]:
#create a column called User Type and assign Existing 
#if User's First Purchase Year Month before the selected Invoice Year Month
tx_uk['UserType'] = 'New'
tx_uk.loc[tx_uk['InvoiceYearMonth'] > tx_uk['FirstPurchaseYearMonth'], 'UserType'] = 'Existing'
tx_uk.loc[tx_uk.UserType=='Existing'].head()

InvoiceNo StockCode                         Description  Quantity  \
297   C543611     82483  WOOD 2 DRAWER CABINET WHITE FINISH        -1   
298   C543611     22632           HAND WARMER RED RETROSPOT        -6   
299   C543611     82483  WOOD 2 DRAWER CABINET WHITE FINISH        -1   
300   C543611     21871                 SAVE THE PLANET MUG        -1   
301   C543611    82494L         WOODEN FRAME ANTIQUE WHITE         -2   

            InvoiceDate  UnitPrice  CustomerID         Country  \
297 2011-02-10 14:38:00       4.95     17850.0  United Kingdom   
298 2011-02-10 14:38:00       1.85     17850.0  United Kingdom   
299 2011-02-10 14:38:00       4.95     17850.0  United Kingdom   
300 2011-02-10 14:38:00       1.06     17850.0  United Kingdom   
301 2011-02-10 14:38:00       2.55     17850.0  United Kingdom   

     InvoiceYearMonth  Revenue       FirstPurchase  FirstPurchaseYearMonth  \
297            201102    -4.95 2010-12-01 08:26:00                  201012   
298            201102   -11.10 2010-12-01 08:26:00                  201012   
299            201102    -4.95 2010-12-01 08:26:00                  201012   
300            201102    -1.06 2010-12-01 08:26:00                  201012   
301            201102    -5.10 2010-12-01 08:26:00                  201012   

     UserType  
297  Existing  
298  Existing  
299  Existing  
300  Existing  
301  Existing

In [133]:
#calculate the Revenue per month for each user type
tx_user_type_revenue = tx_uk.groupby(['InvoiceYearMonth','UserType'])['Revenue'].sum().reset_index()
tx_user_type_revenue.head()

InvoiceYearMonth  UserType    Revenue
0            201012       New  483799.74
1            201101  Existing  195275.51
2            201101       New  156705.77
3            201102  Existing  220994.63
4            201102       New  127859.00

In [134]:
# Exclude 2010 Dec (as it has only new customers) and 2011 Dec (incomplete)
tx_user_type_revenue = tx_user_type_revenue.query("InvoiceYearMonth != 201012 and InvoiceYearMonth != 201112")

plot_data = [
    go.Scatter(
        x=tx_user_type_revenue.query("UserType == 'Existing'")['InvoiceYearMonth'],
        y=tx_user_type_revenue.query("UserType == 'Existing'")['Revenue'],
        name = 'Existing'
    ),
    go.Scatter(
        x=tx_user_type_revenue.query("UserType == 'New'")['InvoiceYearMonth'],
        y=tx_user_type_revenue.query("UserType == 'New'")['Revenue'],
        name = 'New'
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='New vs Existing'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)

Existing customers are showing a positive trend and tell us that our customer base is growing but new customers have a slight negative trend.

Let’s have a better view by looking at the New Customer Ratio:

In [137]:
tx_user_ratio = tx_uk.query(
    "UserType == 'New'").groupby(['InvoiceYearMonth'])['CustomerID'].nunique() / tx_uk.query(
    "UserType == 'Existing'").groupby(['InvoiceYearMonth'])['CustomerID'].nunique() 
tx_user_ratio = tx_user_ratio.reset_index()
tx_user_ratio = tx_user_ratio.dropna()
tx_user_ratio

InvoiceYearMonth  CustomerID
1             201101    1.124224
2             201102    0.904000
3             201103    0.792233
4             201104    0.510166
5             201105    0.343793
6             201106    0.281250
7             201107    0.236589
8             201108    0.192572
9             201109    0.304878
10            201110    0.328852
11            201111    0.236422
12            201112    0.058319

In [138]:
plot_data = [
    go.Bar(
        x=tx_user_ratio.query("InvoiceYearMonth>201101 and InvoiceYearMonth<201112")['InvoiceYearMonth'],
        y=tx_user_ratio.query("InvoiceYearMonth>201101 and InvoiceYearMonth<201112")['CustomerID'],
    )
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='New Customer Ratio'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)

New customer ratio has declined as expected.

### Monthly Retention Rate
Retention rate should be monitored very closely because it indicates how sticky is your service and how well your product fits the market. For making Monthly Retention Rate visualized, we need to calculate how many customers retained from previous month.

Monthly Retention Rate = Retained Customers From Prev. Month/Active Customers Total
We will be using crosstab() function of pandas which makes calculating Retention Rate super easy.

In [139]:
tx_user_purchase = tx_uk.groupby(['CustomerID','InvoiceYearMonth'])['Revenue'].sum().reset_index()
tx_user_purchase

CustomerID  InvoiceYearMonth  Revenue
0         12346.0            201101     0.00
1         12747.0            201012   706.27
2         12747.0            201101   303.04
3         12747.0            201103   310.78
4         12747.0            201105   771.31
...           ...               ...      ...
12325     18283.0            201110   114.65
12326     18283.0            201111   651.56
12327     18283.0            201112   208.00
12328     18287.0            201105   765.28
12329     18287.0            201110  1072.00

[12330 rows x 3 columns]

In [140]:
#create retention matrix with crosstab
tx_retention = pd.crosstab(tx_user_purchase['CustomerID'], tx_user_purchase['InvoiceYearMonth']).reset_index()
tx_retention.head()

InvoiceYearMonth  CustomerID  201012  201101  201102  201103  201104  201105  \
0                    12346.0       0       1       0       0       0       0   
1                    12747.0       1       1       0       1       0       1   
2                    12748.0       1       1       1       1       1       1   
3                    12749.0       0       0       0       0       0       1   
4                    12820.0       0       1       0       0       0       0   

InvoiceYearMonth  201106  201107  201108  201109  201110  201111  201112  
0                      0       0       0       0       0       0       0  
1                      1       0       1       0       1       1       1  
2                      1       1       1       1       1       1       1  
3                      0       0       1       0       0       1       1  
4                      0       0       0       1       1       0       1

In [142]:
#create an array of dictionary which keeps Retained & Total User count for each month
months = tx_retention.columns[2:]
retention_array = []
for i in range(len(months)-1):
    retention_data = {}
    selected_month = months[i+1]
    prev_month = months[i]
    retention_data['InvoiceYearMonth'] = int(selected_month)
    retention_data['TotalUserCount'] = tx_retention[selected_month].sum()
    retention_data['RetainedUserCount'] = tx_retention[(tx_retention[selected_month]>0) & (tx_retention[prev_month]>0)][selected_month].sum()
    retention_array.append(retention_data)
    
retention_array[:2]

[{'InvoiceYearMonth': 201102, 'TotalUserCount': 714, 'RetainedUserCount': 263},
 {'InvoiceYearMonth': 201103, 'TotalUserCount': 923, 'RetainedUserCount': 305}]

In [144]:
#convert the array to dataframe and calculate Retention Rate
tx_retention = pd.DataFrame(retention_array)
tx_retention['RetentionRate'] = tx_retention['RetainedUserCount']/tx_retention['TotalUserCount']
tx_retention

InvoiceYearMonth  TotalUserCount  RetainedUserCount  RetentionRate
0             201102             714                263       0.368347
1             201103             923                305       0.330444
2             201104             817                310       0.379437
3             201105             985                369       0.374619
4             201106             943                417       0.442206
5             201107             899                379       0.421580
6             201108             867                391       0.450980
7             201109            1177                417       0.354291
8             201110            1285                502       0.390661
9             201111            1548                616       0.397933
10            201112             617                402       0.651540

In [145]:
#plot the retention rate graph
plot_data = [
    go.Scatter(
        x=tx_retention.query("InvoiceYearMonth<201112")['InvoiceYearMonth'],
        y=tx_retention.query("InvoiceYearMonth<201112")['RetentionRate'],
        name="organic"
    )
    
]

plot_layout = go.Layout(
        xaxis={"type": "category"},
        title='Monthly Retention Rate'
    )
fig = go.Figure(data=plot_data, layout=plot_layout)
pyoff.iplot(fig)


Monthly Retention Rate significantly jumped from June to August and went back to previous levels afterwards.

### Cohort Based Retention Rate
There is another way of measuring Retention Rate which allows you to see Retention Rate for each cohort. Cohorts are determined as first purchase year-month of the customers. We will be measuring what percentage of the customers retained after their first purchase in each month. This view will help us to see how recent and old cohorts differ regarding retention rate and if recent changes in customer experience affected new customer’s retention or not.

In [147]:
tx_user_purchase.head(2)

CustomerID  InvoiceYearMonth  Revenue
0     12346.0            201101     0.00
1     12747.0            201012   706.27

In [148]:
#create our retention table again with crosstab() and add first purchase year month view
tx_retention = pd.crosstab(tx_user_purchase['CustomerID'], tx_user_purchase['InvoiceYearMonth']).reset_index()
tx_retention = pd.merge(tx_retention,tx_first_purchase[['CustomerID','FirstPurchaseYearMonth']],on='CustomerID')
tx_retention.head()

CustomerID  201012  201101  201102  201103  201104  201105  201106  201107  \
0     12346.0       0       1       0       0       0       0       0       0   
1     12747.0       1       1       0       1       0       1       1       0   
2     12748.0       1       1       1       1       1       1       1       1   
3     12749.0       0       0       0       0       0       1       0       0   
4     12820.0       0       1       0       0       0       0       0       0   

   201108  201109  201110  201111  201112  FirstPurchaseYearMonth  
0       0       0       0       0       0                  201101  
1       1       0       1       1       1                  201012  
2       1       1       1       1       1                  201012  
3       1       0       0       1       1                  201105  
4       0       1       1       0       1                  201101

In [151]:
new_column_names = [ 'm_' + str(column) for column in tx_retention.columns[:-1]]
new_column_names.append('MinPurchaseYearMonth')
tx_retention.columns = new_column_names
tx_retention

m_CustomerID  m_201012  m_201101  m_201102  m_201103  m_201104  \
0          12346.0         0         1         0         0         0   
1          12747.0         1         1         0         1         0   
2          12748.0         1         1         1         1         1   
3          12749.0         0         0         0         0         0   
4          12820.0         0         1         0         0         0   
...            ...       ...       ...       ...       ...       ...   
3945       18280.0         0         0         0         1         0   
3946       18281.0         0         0         0         0         0   
3947       18282.0         0         0         0         0         0   
3948       18283.0         0         1         1         0         1   
3949       18287.0         0         0         0         0         0   

      m_201105  m_201106  m_201107  m_201108  m_201109  m_201110  m_201111  \
0            0         0         0         0         0         0         0   
1            1         1         0         1         0         1         1   
2            1         1         1         1         1         1         1   
3            1         0         0         1         0         0         1   
4            0         0         0         0         1         1         0   
...        ...       ...       ...       ...       ...       ...       ...   
3945         0         0         0         0         0         0         0   
3946         0         1         0         0         0         0         0   
3947         0         0         0         1         0         0         0   
3948         1         1         1         0         1         1         1   
3949         1         0         0         0         0         1         0   

      m_201112  MinPurchaseYearMonth  
0            0                201101  
1            1                201012  
2            1                201012  
3            1                201105  
4            1                201101  
...        ...                   ...  
3945         0                201103  
3946         0                201106  
3947         1                201108  
3948         1                201101  
3949         0                201105  

[3950 rows x 15 columns]

In [152]:
#create the array of Retained users for each cohort monthly
retention_array = []
for i in range(len(months)):
    retention_data = {}
    selected_month = months[i]
    prev_months = months[:i]
    next_months = months[i+1:]
    for prev_month in prev_months:
        retention_data[prev_month] = np.nan
        
    total_user_count = tx_retention[tx_retention.MinPurchaseYearMonth ==  selected_month].MinPurchaseYearMonth.count()
    retention_data['TotalUserCount'] = total_user_count
    retention_data[selected_month] = 1 
    
    query = "MinPurchaseYearMonth == {}".format(selected_month)
    

    for next_month in next_months:
        new_query = query + " and {} > 0".format(str('m_' + str(next_month)))
        retention_data[next_month] = np.round(tx_retention.query(new_query)['m_' + str(next_month)].sum()/total_user_count,2)
    retention_array.append(retention_data)
    
tx_retention = pd.DataFrame(retention_array)
tx_retention.index = months

#showing new cohort based retention table
tx_retention

TotalUserCount  201101  201102  201103  201104  201105  \
InvoiceYearMonth                                                           
201101                       362     1.0    0.23    0.28    0.25    0.34   
201102                       339     NaN    1.00    0.25    0.19    0.28   
201103                       408     NaN     NaN    1.00    0.19    0.26   
201104                       276     NaN     NaN     NaN    1.00    0.22   
201105                       252     NaN     NaN     NaN     NaN    1.00   
201106                       207     NaN     NaN     NaN     NaN     NaN   
201107                       172     NaN     NaN     NaN     NaN     NaN   
201108                       140     NaN     NaN     NaN     NaN     NaN   
201109                       275     NaN     NaN     NaN     NaN     NaN   
201110                       318     NaN     NaN     NaN     NaN     NaN   
201111                       296     NaN     NaN     NaN     NaN     NaN   
201112                        34     NaN     NaN     NaN     NaN     NaN   

                  201106  201107  201108  201109  201110  201111  201112  
InvoiceYearMonth                                                          
201101              0.29    0.26    0.26    0.31    0.35    0.36    0.15  
201102              0.28    0.25    0.26    0.28    0.28    0.31    0.10  
201103              0.22    0.23    0.17    0.26    0.24    0.29    0.09  
201104              0.22    0.22    0.21    0.23    0.23    0.26    0.08  
201105              0.23    0.17    0.17    0.21    0.24    0.27    0.10  
201106              1.00    0.21    0.16    0.25    0.26    0.32    0.10  
201107               NaN    1.00    0.20    0.19    0.23    0.28    0.11  
201108               NaN     NaN    1.00    0.26    0.23    0.26    0.14  
201109               NaN     NaN     NaN    1.00    0.29    0.33    0.12  
201110               NaN     NaN     NaN     NaN    1.00    0.27    0.13  
201111               NaN     NaN     NaN     NaN     NaN    1.00    0.14  
201112               NaN     NaN     NaN     NaN     NaN     NaN    1.00

We can see that first month retention rate became better recently (don’t take Dec ’11 into account) and in almost 1 year, 15% of our customers retain with us.